Wallet:  Add `max_tx_weight` to transaction funding options (take 2)

ismaelsadeeq opened the pull request:
This PR taken over from #29264

The PR added an option `max_tx_weight` to transaction funding RPC's that ensures the resulting transaction weight does not exceed the specified `max_tx_weight` limit.

If `max_tx_weight` is not given `MAX_STANDARD_TX_WEIGHT` is used as the max threshold.

This PR addressed outstanding review comments in #29264

For more context and rationale behind this PR see https://delvingbitcoin.org/t/lightning-transactions-with-v3-and-ephemeral-anchors/418/11?u=instagibbs

ismaelsadeeq committed:
[refactor]: update coin selection algorithms input parameter `max_weight` name

- This commit renames the coin selection algorithms input parameter `max_weight`
  to `max_selection_weight` for clarity.

  The parameter represent the maximum weight of the UTXOs the coin selection algorithm
  should select, not the transaction maximum weight.

- The commit updates the parameter docstring to provide correct description.

- Also updates coin selection unit and fuzzing test variables to match the new name.

ismaelsadeeq committed:
[doc]: update reason for deducting change output weight

`CoinGrinder` will also produce change output, listing all the
Coin selection algorithms that produces change output is not maintainable,
just infer that remaining algorithms all might produce change.

ismaelsadeeq committed:
[wallet, rpc]: add `max_tx_weight` to tx funding options

This allows a transaction's weight to be bound under a certain
weight if possible and desired. This can be beneficial for future
RBF attempts, or whenever a more restricted spend topology is
desired.

Co-authored-by: Greg Sanders <gsanders87@gmail.com>

ismaelsadeeq commented:
Some C.I jobs are failing, putting in draft while I try to address that.

ismaelsadeeq commented:
Note for reviewers: eaf40cea9d31b43b4eb112e561fbadc1911e5732 has been added to ensure the calculation of `max_selection_weight` remains consistent and correct in the next commit 854acd72e99dc9f0924c5f0b5b7a657b39538911
```
int64_t` max_selection_weight = coin_selection_params.m_max_tx_weight.value_or(MAX_STANDARD_TX_WEIGHT) - static_cast<int64_t>(coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
```
This resolves the failure on the CI when compiling with ASan + LSan + UBSan + integer, without dependencies, USDT.
https://cirrus-ci.com/task/6341256662482944

murchandamus commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1374,6 +1374,7 @@ RPCHelpMan sendall()
             RPCResult::Type::OBJ, "", "",
                 {
                     {RPCResult::Type::BOOL, "complete", "If the transaction has a complete set of signatures"},
+                    {RPCResult::Type::NUM, "weight", "The resulting transaction weight in bytes"},
```
In "[rpc]: update `sendall` RPC to return the resulting tx weight" (8e256fe679236710549820ec48ac7c0793fcfd96):

"bytes" is not a unit of weight. I would expect that this is returning the weight in weight units or the vsize in vbytes.

ismaelsadeeq replied:
changes to "The resulting transaction weight in weight units"

murchandamus reviewed (commented):
Concept ACK

ismaelsadeeq commented on src/wallet/coinselection.h:
```diff
@@ -174,10 +174,15 @@ struct CoinSelectionParams {
+    std::optional<int64_t> m_max_tx_weight{std::nullopt};
 
-    CoinSelectionParams(FastRandomContext& rng_fast, size_t change_output_size, size_t change_spend_size,
+    CoinSelectionParams(FastRandomContext& rng_fast, int change_output_size, size_t change_spend_size,
```
Thanks for your review @murchandamus 
> Perhaps if we are getting big into the business of changing the types of various transaction size, vsize and weight variables, we should perhaps consider introducing explicit types for weight, size, and vsize to prevent people accidentally mistaking them for the same type.

I agree with this suggestion and have same concern previously here https://github.com/bitcoin/bitcoin/pull/29242#discussion_r1469871777 should make the code cleaner

Sorry for not explaining the rationale behind this in the commit messages, I will try and explain why I did the change below


On master  we have
```cpp
    int max_inputs_weight = MAX_STANDARD_TX_WEIGHT - (coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
    

   .......
    max_inputs_weight -= (coin_selection_params.change_output_size * WITNESS_SCALE_FACTOR);

```

After this PR maximum transaction weight can be configured by users, so we now subtract from user provided maximum weight or use the default if not given.
```cpp
int max_selection_weight = coin_selection_params.m_max_tx_weight.value_or(MAX_STANDARD_TX
_WEIGHT) - coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR;
```
 
 `m_max_tx_weight` is `int`,` tx_noinputs_size` and `change_output_size`  are both `size_t`.
given that [Bitcoin Core only supports platforms where int is exactly 32 bits](https://github.com/bitcoin/bitcoin/blob/v26.0/src/compat/assumptions.h#L44)
 For signed integers (int), it ranges from -2^(31) to (2^(31)) - 1.
 For unsigned integers (unsigned int), it ranges from 0 to (2^(32)) - 1.
 
The issue arrises when  after compiling with "ASan + LSan + UBSan + integer, without dependencies, USDT" and running the functional test.

From  854acd72e9 we allow passing 0 weight  and test for that in the functional test,  results of the subtracting `coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR` from 0 causes  `max_selection_weight`  to have some positive value instead of 0. hence the test fails see  https://cirrus-ci.com/task/6341256662482944, This is reproducible locally.

Thats why I changed  `change_output_size`,  and `tx_noinputs_size`from `size_t` to `int64_t` and also use same datatype for `m_max_tx_weight` which works fine.



```cpp
int64_t max_selection_weight = coin_selection_params.m_max_tx_weight.value_or(MAX_STANDARD_TX
_WEIGHT) - static_cast<int64_t>(coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
```

But since [Bitcoin Core only supports platforms where int is exactly 32 bits](https://github.com/bitcoin/bitcoin/blob/v26.0/src/compat/assumptions.h#L44) I think using int for `change_output_size`,  and `tx_noinputs_size` should work fine?

ismaelsadeeq replied:
> From https://github.com/bitcoin/bitcoin/commit/854acd72e99dc9f0924c5f0b5b7a657b39538911 we allow passing 0 weight and test for that in the functional test.

After thinking about this more I think it does not make sense to allow passing passing 0 maximum transaction weight, should instead be between minimum transaction weight possible and `MAX_STANDARD_TX_WEIGHT`

ismaelsadeeq replied:
I switched back to status quo, and prevent users from passing `max_tx_weight` below `MIN_STANDARD_TX_NONWITNESS_SIZE * WITNESS_SCALE_FACTOR`

murchandamus replied:
> But since [Bitcoin Core only supports platforms where int is exactly 32 bits](https://github.com/bitcoin/bitcoin/blob/v26.0/src/compat/assumptions.h#L44) I think using int for `change_output_size`, and `tx_noinputs_size` should work fine?

[…]

> I switched back to status quo, and prevent users from passing `max_tx_weight` below `MIN_STANDARD_TX_NONWITNESS_SIZE * WITNESS_SCALE_FACTOR`

I think we may have crossed some wires here. Thanks for the explanation, that was very helpful. I still agree that `size_t` is the _wrong_ type for `change_output_size` and `tx_noinputs_size`. I would support you taking the opportunity because `m_max_tx_weight` is `int` to also change those two to `int`, perhaps even take `change_spend_size` along for that. I would just prefer that it would be better to make it a separate commit and that it should be explained in the commit message.

ismaelsadeeq replied:
I Added a commit that changed them to `int` :)

ismaelsadeeq commented on src/wallet/coinselection.cpp:
```diff
@@ -84,19 +84,19 @@ struct {
  * @param const CAmount& cost_of_change This is the cost of creating and spending a change output.
  *        This plus selection_target is the upper bound of the range.
- * @param int max_selection_weight The maximum allowed weight for a selection result to be valid.
+ * @param int64_t max_selection_weight The maximum allowed weight for a selection result to be valid.
```
Thanks @murchandamus for the link and your review comment, I will look into this and respond as soon as I can.

ismaelsadeeq replied:
You are right `int` is enough, I switched back to the status quo.

Thanks.

ismaelsadeeq commented on src/wallet/spend.cpp:
```diff
@@ -701,7 +701,7 @@ util::Result<SelectionResult> ChooseSelectionResult(interfaces::Chain& chain, co
     }
 
-    // As Knapsack and SRD can create change, also deduce change weight.
+    // Deduce change weight because remaining Coin Selection algorithms can create change output.
```
Fixed thank you!

ismaelsadeeq commented on src/wallet/coinselection.h:
```diff
@@ -174,10 +174,15 @@ struct CoinSelectionParams {
     bool m_include_unsafe_inputs = false;
+    /**
+     * The maximally sized transaction we are willing to make.
+     */
```
Fixed

ismaelsadeeq committed:
[wallet]: update the data type of `change_output_size`, `change_spend_size` and `tx_noinputs_size` to `int`

- This change ensures consistency in transaction size and weight calculation
  within the wallet and prevents conversion overflow when calculating
  `max_selection_weight`.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1050,7 +1050,7 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
         coin_selection_params.change_spend_size = DUMMY_NESTED_P2WPKH_INPUT_SIZE;
     } else {
-        coin_selection_params.change_spend_size = (size_t)change_spend_size;
+        coin_selection_params.change_spend_size = (int)change_spend_size;
```
`change_spend_size` is already an int.

ismaelsadeeq replied:
Yes indeed updated, thanks!

Eunovo commented on test/functional/rpc_psbt.py:
```diff
@@ -187,6 +188,16 @@ def run_test(self):
+        self.log.info("Test that max_tx_weight option is sanity checked and fails when expected")
+        dest_arg = [{self.nodes[0].getnewaddress(): 1}]
+        min_tx_weight = MIN_STANDARD_TX_NONWITNESS_SIZE * WITNESS_SCALE_FACTOR
+        assert_raises_rpc_error(-8, f"Invalid parameter, max tx weight must be between {min_tx_weight} and 400000", self.nodes[0].walletcreatefundedpsbt, [], dest_arg, 0, {"max_tx_weight": -1})
```
I can see that the "400000" here is the `MAX_STANDARD_TX_WEIGHT` but I think it's better to define the `MAX_STANDARD_TX_WEIGHT` constant and use it here. This makes it obvious to anyone reading what "400000" is.

ismaelsadeeq replied:
Yes can also be used in future tests, thinking about defining it in 
`test_framework/blocktools.py`

ismaelsadeeq replied:
Fixed thanks

Eunovo commented on src/wallet/rpc/spend.cpp:
```diff
@@ -787,6 +796,8 @@ RPCHelpMan fundrawtransaction()
                                     },
                                 },
                              },
+                            {"max_tx_weight", RPCArg::Type::NUM, RPCArg::Default{MAX_STANDARD_TX_WEIGHT}, "The maximum acceptable transaction weight.\n"
```
Nit: might be good to define this in one place since you use it multiple times.

ismaelsadeeq replied:
What should be defined?
The help text ? not sure if this is a good idea though so leaving this as is!

ismaelsadeeq commented:
Force pushed from https://github.com/bitcoin/bitcoin/commit/53ee67de7e388703cb672e67ffaecdde53eb01de to https://github.com/bitcoin/bitcoin/pull/29523/commits/3a64b3a339fe55c2642369b00a382231871c2838 [compare diff](https://github.com/bitcoin/bitcoin/compare/53ee67de7e388703cb672e67ffaecdde53eb01de..3a64b3a339fe55c2642369b00a382231871c2838)

- Addressed [review comment ](https://github.com/bitcoin/bitcoin/pull/29523#discussion_r1542667882)
- Rebased on master after #29788

furszy reviewed (commented):
Any chance to turn the first commit into an scripted-diff?

Eunovo commented:
Tested ACK https://github.com/bitcoin/bitcoin/pull/29523/commits/3a64b3a339fe55c2642369b00a382231871c283

ismaelsadeeq commented:
> Any chance to turn the first commit into an scripted-diff?

The commit message was misleading It's not just renaming `max_weight`, it also updates the input parameter docstring to match match the new name.
And also the fuzz test renames has more info like `low_max_selection_weight` and `high_max_selection_weight`.

I've updated the commit message of the first commit https://github.com/bitcoin/bitcoin/pull/29523/commits/270639324a6637e0025e1458221ab488a8a09554

ismaelsadeeq commented:
Rebased and force pushed from 3d76c37106a4f80fd2ac410696b1049c3aa321c5 to b24b7a9a6a1a10024c639e9d4ee27aa4d0ce653e to fix C.I  failure [compare diff](https://github.com/bitcoin/bitcoin/compare/3d76c37106a4f80fd2ac410696b1049c3aa321c5..b24b7a9a6a1a10024c639e9d4ee27aa4d0ce653e)

ismaelsadeeq commented:
Rebased to fix conflict after https://github.com/bitcoin/bitcoin/pull/28366  was merged.
[b24b7a9a6..845d16c1b](https://github.com/bitcoin/bitcoin/compare/b24b7a9a6a1a10024c639e9d4ee27aa4d0ce653e..845d16c1bbffb574b22c79ce08e2b6eba31865bc)

furszy commented on src/wallet/spend.cpp:
```diff
@@ -692,7 +692,9 @@ util::Result<SelectionResult> ChooseSelectionResult(interfaces::Chain& chain, co
 
     // Maximum allowed weight for selected coins.
-    int max_selection_weight = MAX_STANDARD_TX_WEIGHT - (coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR);
+    int max_transaction_weight = coin_selection_params.m_max_tx_weight.value_or(MAX_STANDARD_TX_WEIGHT);
+    int tx_weight_no_input = coin_selection_params.tx_noinputs_size * WITNESS_SCALE_FACTOR;
+    int max_selection_weight = std::max((max_transaction_weight - tx_weight_no_input), 0);
```
In d9ec6e7f28f0:

Instead of setting it to 0, could error out early if `max_selection_weight <= 0` (a test for this would be nice).

ismaelsadeeq replied:
Thanks fixed.

ismaelsadeeq replied:
Also added a test for this

furszy commented on src/wallet/spend.cpp:
```diff
@@ -702,7 +704,8 @@ util::Result<SelectionResult> ChooseSelectionResult(interfaces::Chain& chain, co
 
     // Deduct change weight because remaining Coin Selection algorithms can create change output
-    max_selection_weight -= (coin_selection_params.change_output_size * WITNESS_SCALE_FACTOR);
+    int change_outputs_weight = coin_selection_params.change_output_size * WITNESS_SCALE_FACTOR;
+    max_selection_weight = std::max((max_selection_weight - change_outputs_weight), 0);
```
Same as above, would be nice to error out early if `max_selection_weight <= 0`. No need to run the selection algorithms under such circumstance.

ismaelsadeeq replied:
Fixed, and also add a test case for it.

furszy commented on test/functional/test_framework/blocktools.py:
```diff
@@ -48,6 +48,7 @@
 
 MAX_BLOCK_SIGOPS = 20000
 MAX_BLOCK_SIGOPS_WEIGHT = MAX_BLOCK_SIGOPS * WITNESS_SCALE_FACTOR
+MAX_STANDARD_TX_WEIGHT = 400000
```
nit: could replace [this line](https://github.com/bitcoin/bitcoin/blob/2d21060af831fa8f8f1791b4464961d5f28a88cb/test/functional/p2p_segwit.py#L1265) with this constant.

ismaelsadeeq replied:
leaving this as is, as I suspect there might be other places that needs to be replaced also.
this and others can come in a separate PR ?

furszy reviewed (commented):
few small comments

ismaelsadeeq commented:
thanks for your review @furszy will force push to address this comments shortly.

rkrux reviewed (approved):
Concept ACK [a120fc7](https://github.com/bitcoin/bitcoin/pull/29523/commits/a120fc7caf863aa02ca27d6eae52f3b35202e01d), but I would look at [723a8da](https://github.com/bitcoin/bitcoin/pull/29523/commits/723a8da8e562451facc8daa5d6ba61977f68cafe) commit a bit more again; `make, test/functional` are successful though.

Following transaction funding RPCs are affected with the addition of `max_tx_weight` option:
  1. `fundrawtransaction`
  2. `walletcreatefundedpsbt`
  3. `send`

ismaelsadeeq commented on src/wallet/coinselection.h:
```diff
@@ -139,9 +139,9 @@ struct CoinSelectionParams {
     FastRandomContext& rng_fast;
     /** Size of a change output in bytes, determined by the output type. */
-    size_t change_output_size = 0;
+    int change_output_size = 0;
```
It's simply just to ensure consistency in transaction size and weight calculation within the wallet and prevents conversion overflow when calculating `max_selection_weight` as it happened here https://cirrus-ci.com/task/6341256662482944

---
I think it definitely will be worth it to define a datatype for size in the entire codebase.
But I am limiting this change to affects only the wallet size variables `change_output_size`, `change_spend_size` and `tx_noinputs_size`, because they relate to this PR.

ismaelsadeeq commented on test/functional/rpc_psbt.py:
```diff
@@ -186,6 +191,49 @@ def run_test(self):
+        max_tx_weight_sufficient = 1000 # 1k vbytes is enough
+        psbt = self.nodes[0].walletcreatefundedpsbt(outputs=dest_arg,locktime=0, options={"max_tx_weight": max_tx_weight_sufficient})["psbt"]
+        weight = self.nodes[0].decodepsbt(psbt)["tx"]["weight"]
+        # ensure the transactions weight is below the transaction weight.
```
Fixed.

ismaelsadeeq commented on test/functional/rpc_psbt.py:
```diff
@@ -186,6 +191,49 @@ def run_test(self):
+        # min_tx_weight is greater than transaction weight without inputs
+        assert_greater_than(min_tx_weight, tx_weight_without_inputs)
+
+        # In other to test for when the passed max weight is less than the transaction weight without inputs
```
Yes updated

furszy commented on src/wallet/spend.cpp:
```diff
@@ -706,7 +716,11 @@ util::Result<SelectionResult> ChooseSelectionResult(interfaces::Chain& chain, co
     // Deduct change weight because remaining Coin Selection algorithms can create change output
-    max_selection_weight -= (coin_selection_params.change_output_size * WITNESS_SCALE_FACTOR);
+    int change_outputs_weight = coin_selection_params.change_output_size * WITNESS_SCALE_FACTOR;
+    max_selection_weight = max_selection_weight - change_outputs_weight;
```
tiny nit:
```suggestion
    max_selection_weight -= change_outputs_weight;
```

ismaelsadeeq replied:
fixed

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1091,13 +1106,16 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     // Static vsize overhead + outputs vsize. 4 version, 4 nLocktime, 1 input count, 1 witness overhead (dummy, flag, stack size)
     coin_selection_params.tx_noinputs_size = 10 + GetSizeOfCompactSize(vecSend.size()); // bytes for output count
 
+    LogInfo("input count size %s \n", GetSizeOfCompactSize(vecSend.size()));
```
guess that this was used for testing purposes and you forgot to remove it?

ismaelsadeeq replied:
Yes, removed

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1091,13 +1106,16 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
         CTxOut txout(recipient.nAmount, GetScriptForDestination(recipient.dest));
 
         // Include the fee cost for outputs.
         coin_selection_params.tx_noinputs_size += ::GetSerializeSize(txout);
+        LogInfo("Output size %s \n", static_cast<int>(::GetSerializeSize(txout)));
```
Same as above, guess that this was used for testing purposes and you forgot to remove it?

ismaelsadeeq replied:
removed

furszy commented on test/functional/rpc_psbt.py:
```diff
@@ -186,6 +191,49 @@ def run_test(self):
+
+        # Test for max_tx_weight just enough to include inputs but not change output
+        assert_raises_rpc_error(-4, "Maximum transaction weight is too low, can not accommodate change output", self.nodes[0].walletcreatefundedpsbt, [], dest_arg_large, 0, {"max_tx_weight": (large_tx_vsize_without_inputs + 1) * WITNESS_SCALE_FACTOR})
+        # Test that it has to accommodate a change output
+        large_tx_vsize_with_change = large_tx_vsize_without_inputs + p2wpkh_output_vsize
+        assert_raises_rpc_error(-4, "Maximum transaction weight is too low, can not accommodate change output", self.nodes[0].walletcreatefundedpsbt, [], dest_arg_large, 0, {"max_tx_weight": (large_tx_vsize_with_change) * WITNESS_SCALE_FACTOR})
```
Isn't the purpose of this test case to include the change output and fail after wise? (not with the "can not accommodate change output" error message)

ismaelsadeeq replied:
Yes, good catch.
Updated

furszy commented on test/functional/rpc_psbt.py:
```diff
@@ -985,6 +1033,5 @@ def test_psbt_input_keys(psbt_input, keys):
         self.log.info("Test descriptorprocesspsbt raises if an invalid sighashtype is passed")
         assert_raises_rpc_error(-8, "'all' is not a valid sighash parameter.", self.nodes[2].descriptorprocesspsbt, psbt, [descriptor], sighashtype="all")
 
-
```
`PEP 8: E305 expected 2 blank lines after class or function definition, found 1`. Let's not remove this line.

ismaelsadeeq replied:
added the line back, thanks

furszy commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1564,8 +1565,9 @@ RPCHelpMan sendall()
-
-            return FinishTransaction(pwallet, options, rawTx);
+            auto result = FinishTransaction(pwallet, options, rawTx);
+            result.pushKV("weight", tx_size.weight);
```
Not sure about using `tx_size`, we would be returning an estimation of the final size of tx. Not the real/current tx weight.

What about getting the current tx weight by serializing the `CMutableTransaction` object that appears inside `FinishTransaction` and explain in the RPC return value description that the returned weight will not be the final one if the transaction is not complete.

rkrux replied:
Interesting, I agree with this point. Returning the exact size of the transaction would make the RPC definition more robust.

rkrux replied:
This commit (along with this^ suggestion) [b3ac117](https://github.com/bitcoin/bitcoin/pull/29523/commits/b3ac1179ff90fe261af4e6dc9c7af64e1518b435) can be a separate PR as well, doesn't seem necessary for this PR to be merged if I didn't miss anything.

ismaelsadeeq replied:
Yes it's not needed.
I removed this commit
This can come as a separate PR, the size output can also be added to the other funding RPC's?

furszy reviewed (commented):
Code reviewed b3ac1179ff90f, getting closer. Left few topics.

ismaelsadeeq commented:
Thanks for the reviews @furszy @rkrux 

Force-pushed from b3ac1179ff90fe261af4e6dc9c7af64e1518b435 to d8febf6d9ea018cf8e980ee036fb5ccd8ea8887a
Compare diff [b3ac1179ff..734076c6d](https://github.com/bitcoin/bitcoin/compare/b3ac1179ff90fe261af4e6dc9c7af64e1518b435..734076c6de1781f957c8bc3bf7ed6951920cfcf6)

---

Changes:
- Rebased after [PR 30309](https://github.com/bitcoin/bitcoin/pull/30309) and updated the maximum transaction weight check to use the user passed value when given or default `MAX_STANDARD_TX_WEIGHT`.
- Removed miscellaneous logs.
- Added back the removed line in the `rpc_psbt.py` functional test according PEP guidelines.
- Removed the last commit that returned the estimated size in the `send` RPC based on the review [here](https://github.com/bitcoin/bitcoin/pull/29523#discussion_r1655511363).

furszy commented on src/wallet/spend.cpp:
```diff
@@ -1002,7 +1011,11 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
-
+    coin_selection_params.m_max_tx_weight = coin_control.m_max_tx_weight.value_or(MAX_STANDARD_TX_WEIGHT);
+    int minimum_tx_weight = MIN_STANDARD_TX_NONWITNESS_SIZE * WITNESS_SCALE_FACTOR;
+    if (coin_selection_params.m_max_tx_weight.value() < minimum_tx_weight || coin_selection_params.m_max_tx_weight.value() > MAX_STANDARD_TX_WEIGHT) {
```
In 734076c6de178:

nano nit: shorter
```suggestion
    if (*coin_selection_params.m_max_tx_weight < minimum_tx_weight || *coin_selection_params.m_max_tx_weight > MAX_STANDARD_TX_WEIGHT) {
```

furszy reviewed (commented):
utACK 734076c6de1781f957c8bc3bf7ed6951920cfcf6

rkrux reviewed (approved):
reACK [734076c](https://github.com/bitcoin/bitcoin/pull/29523/commits/734076c6de1781f957c8bc3bf7ed6951920cfcf6)

Ran make and functional tests, all successful.

@ismaelsadeeq This [diff comparison b3ac117..734076c](https://github.com/bitcoin/bitcoin/compare/b3ac1179ff90fe261af4e6dc9c7af64e1518b435..734076c6de1781f957c8bc3bf7ed6951920cfcf6) contains a lot of unrelated changes though.

ismaelsadeeq commented:
> @ismaelsadeeq This [diff comparison b3ac117..734076c](https://github.com/bitcoin/bitcoin/compare/b3ac1179ff90fe261af4e6dc9c7af64e1518b435..734076c6de1781f957c8bc3bf7ed6951920cfcf6) contains a lot of unrelated changes though.

Yes that's because of the recent rebase.

achow101 commented:
ACK 734076c6de1781f957c8bc3bf7ed6951920cfcf6
